File: src/topols/embedding/ports.py
"""Input ports of a block (`auto_ports`), lifting a finished layer to a common
ceiling and sealing output ends (`ceiling`, `seal_brute_frontier`), and the
space-time volume of an embedding.
"""

import math
import os


def _ceil_dbg(tag, key, extra="-"):
    """Append a line to the file named by $TOPOLS_H_DEBUG, if set (debug aid)."""
    _p = os.environ.get("TOPOLS_H_DEBUG")
    if _p:
        with open(_p, "a") as _fh:
            _fh.write(f"{tag}\t{key}\t{extra}\n")

from topols.routing.color_algebra import ORI_MAP, edge_tracer
from topols.routing.boundary import vertical_z_path

# ---------------------------------------------------------------------------
# Other Function
# ---------------------------------------------------------------------------

def calculate_space_time(pos, paths, x_min_floor, x_max_floor, y_min_floor, y_max_floor):
    """Extents and space-time volume of an embedding.

    x/y extents are the footprint fixed by the floor bounds (inclusive);
    the z extent is measured from the cubes and paths. Returns
    `(x_length, y_length, z_length, x_length * y_length * z_length)`.
    """
    all_points = [coord for coord in pos.values()]
    all_points += [pt for path in paths for pt in path]
    _, _, zs = zip(*all_points)
    z_min, z_max = min(zs), max(zs)
    x_length = x_max_floor - x_min_floor + 1
    y_length = y_max_floor - y_min_floor + 1
    z_length = z_max - z_min
    volume = x_length * y_length * z_length
    return x_length, y_length, z_length, volume


def auto_ports(num_qubits, z_level=0, edge_dist=2, length=2):
    """Input ports of the first block, laid out on a 2D grid.

    Qubit `i` gets a cube at `z_level`; `length` qubits per row, rows in
    serpentine order, `edge_dist` cells between neighbours so that a pipe
    fits in between. Every port starts with orientation `"i"` and type 0.

    Returns:
        `(positions, orientations, types)`, each keyed by qubit index.
    """
    if length is None:
        length = math.ceil(num_qubits ** 0.5)
    width = math.ceil(num_qubits / length)

    input_port_loc = {}
    input_port_ori = {}
    input_port_type = {}

    idx = 0
    for j in range(width):
        i_s = range(length) if j % 2 == 0 else reversed(range(length))
        for i in i_s:
            if idx >= num_qubits:
                break
            input_port_loc[idx] = (i * edge_dist, j * edge_dist, z_level)
            input_port_ori[idx] = 'i'
            input_port_type[idx] = 0
            idx += 1
        if idx >= num_qubits:
            break

    return input_port_loc, input_port_ori, input_port_type


def ceiling(best_state, ceiling_track, node_type, final=False):
    """Commit the lift of a finished layer to a common ceiling plane.

    `EmbeddingState.reward` computes `ceiling_track`: for every node of
    the layer that stays open, the vertical path to the ceiling and the
    orientation/type its end must have. This function applies it to
    `best_state`: the layer's nodes are renamed `<node>_old` and the
    lifted ends become the new `<node>` entries (the next layer's input
    ports), paths and occupancy are extended, and idle ends are recorded
    in `idle_place` / `idle_h_track` so that their chain still knows the
    real node it started from.

    With `final=True` the lift is the compile's last step: every end is
    given a definite colour (type 0), including the flip for a Hadamard
    that sits between the last real node and the output port
    (`HTable.needs_flip_to_end`). Without it, idle ends stay idle and the
    next real node applies any Hadamard.

    Args:
        best_state: `EmbeddingState` to update (mutated and returned).
        ceiling_track: `{node: {"path": [...], "ori": ..., "type": ...}}`.
        node_type: `{node: type}` of the layer being lifted.
        final: True for the final seal of the compile.
    """
    # ORI_MAP: topols.routing.color_algebra.ORI_MAP (module-level import).
    ori_map = ORI_MAP

    occ = set(best_state.occupied)

    for key in list(best_state.embed_node_pos.keys()):
        if key in node_type:
            best_state.embed_node_pos[f"{key}_old"] = best_state.embed_node_pos[key]
            del best_state.embed_node_pos[key]

    for key in list(best_state.embed_node_ori.keys()):
        if key in node_type:
            best_state.embed_node_ori[f"{key}_old"] = best_state.embed_node_ori[key]
            del best_state.embed_node_ori[key]

    for key in list(best_state.embed_node_type.keys()):
        if key in node_type:
            best_state.embed_node_type[f"{key}_old"] = best_state.embed_node_type[key]
            del best_state.embed_node_type[key]

    for key in list(best_state.t_track.keys()):
        if key in node_type:
            best_state.t_track[f"{key}_old"] = best_state.t_track[key]
            del best_state.t_track[key]


    for key, dic in ceiling_track.items():
        path = dic["path"]
        best_state.embed_node_pos[key] = path[-1]

    for key, dic in ceiling_track.items():
        if "ori" in dic:
            ori = dic["ori"]
            best_state.embed_node_ori[key] = ori

    if final:
        # A real node whose lift ends at the output port: if the wire from
        # the node to the port carries an odd number of Hadamards, the end
        # colour computed by reward() must flip. Recompute its formula with
        # the flipped type. Idle chains are handled in the branch below.
        for key, dic in ceiling_track.items():
            t0 = node_type.get(key)
            if t0 not in (0, 1, 4, 5) or "ori" not in dic:
                continue
            old = f"{key}_old"
            if old not in best_state.embed_node_ori or not best_state.hadamard_edges.needs_flip_to_end(old):
                continue
            path = tuple(dic["path"])
            if t0 in (4, 5):
                curr_type, last_dir = edge_tracer(path, (best_state.embed_node_ori[old], 0))
                curr_type = 1 - curr_type
                if ori_map[(last_dir, curr_type, 0)] == 'k':
                    dic["ori"] = ori_map[(last_dir, curr_type, 1)]
                    dic["type"] = 1
                else:
                    dic["ori"] = ori_map[(last_dir, curr_type, 0)]
                    dic["type"] = t0
            else:
                curr_type, last_dir = edge_tracer(path, (best_state.embed_node_ori[old], t0))
                curr_type = 1 - curr_type
                if ori_map[(last_dir, curr_type, t0)] == 'k':
                    dic["ori"] = ori_map[(last_dir, curr_type, 1 if t0 != 1 else 0)]
                    dic["type"] = 1 if t0 != 1 else 0
                else:
                    dic["ori"] = ori_map[(last_dir, curr_type, t0)]
                    dic["type"] = t0
            best_state.embed_node_ori[key] = dic["ori"]

    for key, dic in ceiling_track.items():
        type = dic["type"]
        if type > 1 and type not in (4, 5):
            if final:
                type = 0
                path = dic["path"]
                start_node, cur_path, h_count = best_state.idle_h_track[key]
                best_state.idle_h_track[key] = [
                            start_node,
                            tuple(list(path)[::-1] + list(cur_path)[1:]),
                            h_count
                            ]
                tol_path = list(path)[::-1] + list(cur_path)[1:]
                if best_state.embed_node_type[start_node] in (4, 5):
                    curr_type, last_dir = edge_tracer(tuple(tol_path)[::-1], (best_state.embed_node_ori[start_node], 0))
                else:
                    curr_type, last_dir = edge_tracer(tuple(tol_path)[::-1], (best_state.embed_node_ori[start_node], best_state.embed_node_type[start_node]))
                # The chain runs from its origin to the output port, so it
                # owns every Hadamard after the origin.
                if best_state.hadamard_edges.needs_flip_to_end(start_node):
                    curr_type = 1 - curr_type
                ori = ori_map[(last_dir, curr_type, 0)]
                best_state.embed_node_ori[key] = ori
            else:
                type = 2
                path = dic["path"]
                if key in best_state.idle_h_track:
                    start_node, cur_path, h_count = best_state.idle_h_track[key]
                    _ceil_dbg("ceiling_keep", key, h_count)
                    best_state.idle_h_track[key] = [
                                start_node,
                                tuple(list(path)[::-1] + list(cur_path)[1:]),
                                h_count
                                ]
                else:
                    _ceil_dbg("ceiling_fresh_hcount0", key)
                    best_state.idle_h_track[key] = [
                                f"{key}_old",
                                tuple(list(path)[::-1]),
                                0
                                ]
        elif type in (4, 5):
            type = 0
        best_state.embed_node_type[key] = type

    ceiling_paths = [dic["path"] for dic in ceiling_track.values()]
    best_state.embed_path = tuple(list(best_state.embed_path)+ceiling_paths)

    for dic in ceiling_track.values():
        path = dic["path"]
        for pt in path:
            occ.add(pt)
    best_state.occupied = frozenset(occ)

    idle_place = {}
    for key, dic in ceiling_track.items():
        type = dic["type"]
        if type > 1 and type not in (4, 5):
            coord = dic["path"][-1]
            idle_place[key] = coord
    best_state.idle_place = idle_place

    return best_state


def seal_brute_frontier(best_state):
    """Final seal for a frontier produced by basic_embedding.

    basic_embedding places every real node of the layer as `X_old` and
    leaves an idle stub `X` (type 2) directly above it at the ceiling, with
    `idle_h_track[X] = [X_old, path, ...]`; H boxes it keeps are type-3
    stubs on such a chain. ceiling(final=True) cannot be used on that
    state: it renames every layer key to `X_old`, which would overwrite the
    real node's entry, and reward() cannot build its ceiling_track because
    the stubs carry no orientation. The stubs are already at the top, so
    the seal is only the colour step of ceiling(final=True)'s idle branch:
    trace the chain from its origin, flip if the wire from the origin to
    the output port carries an odd number of Hadamards, set type 0.
    Mutates and returns best_state.
    """
    ori_map = ORI_MAP
    frontier = [key for key, (start_node, _p, _h) in best_state.idle_h_track.items()
                if best_state.embed_node_type.get(key) in (2, 3)
                and start_node in best_state.embed_node_ori]
    if not frontier:
        return best_state

    # Like ceiling(), bring every output end to one ceiling: basic_embedding
    # stacks layer by layer, so a qubit whose last layer held a box may end
    # higher than the others. Extend each lower stub straight up when the
    # column is free; the extension belongs to the same idle chain, so it
    # is prepended to the chain path (which runs from the chain end back
    # to its origin).
    occ = set(best_state.occupied)
    z_top = max(best_state.embed_node_pos[k][2] for k in frontier)
    paths = list(best_state.embed_path)
    for key in frontier:
        x, y, z = best_state.embed_node_pos[key]
        if z >= z_top:
            continue
        column = [(x, y, zz) for zz in range(z + 1, z_top + 1)]
        if any(pt in occ for pt in column):
            continue
        seg = vertical_z_path((x, y, z), (x, y, z_top))
        start_node, cur_path, h = best_state.idle_h_track[key]
        best_state.idle_h_track[key] = [start_node, tuple(seg[::-1]) + tuple(cur_path)[1:], h]
        best_state.embed_node_pos[key] = (x, y, z_top)
        if key in best_state.idle_place:
            best_state.idle_place[key] = (x, y, z_top)
        # The stub's old position is no longer a node, so the recorded path
        # that ends there (basic_embedding's `X_old -> X` vertical) must be
        # extended in place rather than joined by a second segment --
        # export's get_edge() keys every path by the nodes at its two ends
        # and a junction with no node there is a KeyError downstream.
        old_end = (x, y, z)
        extended = False
        for idx, p in enumerate(paths):
            pp = [tuple(int(c) for c in q) for q in p]
            if pp[-1] == old_end:
                paths[idx] = tuple(pp + seg[1:]); extended = True; break
            if pp[0] == old_end:
                paths[idx] = tuple(list(reversed(seg[1:])) + pp); extended = True; break
        if not extended:
            paths.append(tuple(seg))
        occ.update(seg)
    best_state.embed_path = tuple(paths)
    best_state.occupied = frozenset(occ)

    for key in frontier:
        start_node, cur_path, _h = best_state.idle_h_track[key]
        st = 0 if best_state.embed_node_type.get(start_node) in (4, 5) else best_state.embed_node_type.get(start_node, 0)
        curr_type, last_dir = edge_tracer(tuple(cur_path)[::-1], (best_state.embed_node_ori[start_node], st))
        if best_state.hadamard_edges.needs_flip_to_end(start_node):
            curr_type = 1 - curr_type
        best_state.embed_node_ori[key] = ori_map[(last_dir, curr_type, 0)]
        best_state.embed_node_type[key] = 0
    return best_state

File: rust/topols_core/src/embedding/ports.rs
//! Input-port layout, the lift of a finished layer to a common ceiling and
//! the final seals (ports.py).

use indexmap::IndexMap;

use crate::embedding::node::{is_cube, trace_type, NodeId, NodeType};
use crate::embedding::state::{CeilingEntry, EmbeddingState, NodeMap, Path, Track};
use crate::geometry::{Cell, Floors};
use crate::routing::astar::Occ;
use crate::routing::boundary::vertical_z_path;
use crate::routing::color::{edge_tracer, ori_map, Axis};

/// `auto_ports`: qubit i -> cell on a serpentine grid at `z_level`, `length`
/// per row, `edge_dist` apart. Orientation 'i' and type 0 for every port.
pub fn auto_ports(num_qubits: usize, z_level: i32, edge_dist: i32, length: Option<usize>) -> IndexMap<usize, Cell> {
    let length = length.unwrap_or_else(|| (num_qubits as f64).sqrt().ceil() as usize);
    let width = (num_qubits + length - 1) / length;
    let mut loc = IndexMap::new();
    let mut idx = 0usize;
    'rows: for j in 0..width {
        let row: Vec<usize> = if j % 2 == 0 { (0..length).collect() } else { (0..length).rev().collect() };
        for i in row {
            if idx >= num_qubits {
                break 'rows;
            }
            loc.insert(idx, Cell::new((i as i32) * edge_dist, (j as i32) * edge_dist, z_level));
            idx += 1;
        }
        if idx >= num_qubits {
            break;
        }
    }
    loc
}

pub const PORT_ORI: Axis = Axis::I;

/// `calculate_space_time`: (x_length, y_length, z_length, volume).
pub fn calculate_space_time(pos: &NodeMap<Cell>, paths: &[Path], floors: &Floors) -> (f64, f64, f64, f64) {
    let zs = pos.values().map(|c| c.z).chain(paths.iter().flat_map(|p| p.iter().map(|c| c.z)));
    let (mut zmin, mut zmax) = (i32::MAX, i32::MIN);
    for z in zs {
        zmin = zmin.min(z);
        zmax = zmax.max(z);
    }
    let x_length = floors.x_max.unwrap() - floors.x_min.unwrap() + 1.0;
    let y_length = floors.y_max.unwrap() - floors.y_min.unwrap() + 1.0;
    let z_length = (zmax - zmin) as f64;
    (x_length, y_length, z_length, x_length * y_length * z_length)
}

fn rename_old<V: Clone>(map: &mut NodeMap<V>, node_type: &NodeMap<NodeType>) {
    let keys: Vec<NodeId> = map.keys().copied().collect();
    for key in keys {
        if node_type.contains_key(&key) {
            let v = map[&key].clone();
            map.insert(key.as_old(), v);
            map.shift_remove(&key);
        }
    }
}

fn reversed(p: &[Cell]) -> Path {
    p.iter().rev().copied().collect()
}

/// `ceiling`: commit the lift described by `ceiling_track` to `state`
/// (mutated and returned). With `final_seal`, colour every output end.
pub fn ceiling(state: &mut EmbeddingState, ceiling_track: &mut NodeMap<CeilingEntry>, node_type: &NodeMap<NodeType>, final_seal: bool) {
    let mut occ: Occ = state.occupied.clone();

    rename_old(&mut state.pos, node_type);
    rename_old(&mut state.ori, node_type);
    rename_old(&mut state.typ, node_type);
    rename_old(&mut state.t_track, node_type);

    for (key, e) in ceiling_track.iter() {
        state.pos.insert(*key, *e.path.last().unwrap());
    }
    for (key, e) in ceiling_track.iter() {
        if let Some(o) = e.ori {
            state.ori.insert(*key, o);
        }
    }

    if final_seal {
        // A real node whose lift ends at the output port: if the wire from the
        // node to the port carries an odd number of Hadamards, the end colour
        // computed by reward() must flip.
        for (key, e) in ceiling_track.iter_mut() {
            let Some(&t0) = node_type.get(key) else { continue };
            if !is_cube(t0) || e.ori.is_none() {
                continue;
            }
            let old = key.as_old();
            if !state.ori.contains_key(&old) || !state.htable.needs_flip_to_end(old) {
                continue;
            }
            if t0 == 4 || t0 == 5 {
                let (ct, ld) = edge_tracer(&e.path, state.ori[&old], 0);
                let ct = 1 - ct;
                if ori_map(ld, ct, 0) == Axis::K {
                    e.ori = Some(ori_map(ld, ct, 1));
                    e.typ = 1;
                } else {
                    e.ori = Some(ori_map(ld, ct, 0));
                    e.typ = t0;
                }
            } else {
                let (ct, ld) = edge_tracer(&e.path, state.ori[&old], t0 as u8);
                let ct = 1 - ct;
                if ori_map(ld, ct, t0 as u8) == Axis::K {
                    let t2: u8 = if t0 != 1 { 1 } else { 0 };
                    e.ori = Some(ori_map(ld, ct, t2));
                    e.typ = t2 as NodeType;
                } else {
                    e.ori = Some(ori_map(ld, ct, t0 as u8));
                    e.typ = t0;
                }
            }
            state.ori.insert(*key, e.ori.unwrap());
        }
    }

    for (key, e) in ceiling_track.iter() {
        let mut t = e.typ;
        if t > 1 && t != 4 && t != 5 {
            if final_seal {
                t = 0;
                let tr = state.idle_h_track.get(key).expect("chain end without track").clone();
                let mut tol: Path = reversed(&e.path);
                tol.extend_from_slice(&tr.path[1..]);
                state.idle_h_track.insert(*key, Track { origin: tr.origin, path: tol.clone(), h: tr.h });
                let (mut ct, ld) = edge_tracer(&reversed(&tol), state.ori[&tr.origin], trace_type(state.typ[&tr.origin]));
                if state.htable.needs_flip_to_end(tr.origin) {
                    ct = 1 - ct;
                }
                state.ori.insert(*key, ori_map(ld, ct, 0));
            } else {
                t = 2;
                if let Some(tr) = state.idle_h_track.get(key).cloned() {
                    let mut p: Path = reversed(&e.path);
                    p.extend_from_slice(&tr.path[1..]);
                    state.idle_h_track.insert(*key, Track { origin: tr.origin, path: p, h: tr.h });
                } else {
                    state.idle_h_track.insert(*key, Track { origin: key.as_old(), path: reversed(&e.path), h: 0 });
                }
            }
        } else if t == 4 || t == 5 {
            t = 0;
        }
        state.typ.insert(*key, t);
    }

    for e in ceiling_track.values() {
        state.paths.push(e.path.clone());
        occ.extend(e.path.iter().copied());
    }
    state.occupied = occ;

    let mut idle_place = NodeMap::new();
    for (key, e) in ceiling_track.iter() {
        if e.typ > 1 && e.typ != 4 && e.typ != 5 {
            idle_place.insert(*key, *e.path.last().unwrap());
        }
    }
    state.idle_place = idle_place;
}

/// `seal_brute_frontier`: final seal of a frontier left by `basic_embedding`.
pub fn seal_brute_frontier(state: &mut EmbeddingState) {
    let frontier: Vec<NodeId> = state
        .idle_h_track
        .iter()
        .filter(|(k, tr)| matches!(state.typ.get(*k), Some(2) | Some(3)) && state.ori.contains_key(&tr.origin))
        .map(|(k, _)| *k)
        .collect();
    if frontier.is_empty() {
        return;
    }
    let mut occ = state.occupied.clone();
    let z_top = frontier.iter().map(|k| state.pos[k].z).max().unwrap();
    let mut paths = state.paths.clone();
    for &key in &frontier {
        let Cell { x, y, z } = state.pos[&key];
        if z >= z_top {
            continue;
        }
        if (z + 1..=z_top).any(|zz| occ.contains(&Cell::new(x, y, zz))) {
            continue;
        }
        let seg = vertical_z_path(Cell::new(x, y, z), Cell::new(x, y, z_top));
        let tr = state.idle_h_track[&key].clone();
        let mut p: Path = reversed(&seg);
        p.extend_from_slice(&tr.path[1..]);
        state.idle_h_track.insert(key, Track { origin: tr.origin, path: p, h: tr.h });
        state.pos.insert(key, Cell::new(x, y, z_top));
        if state.idle_place.contains_key(&key) {
            state.idle_place.insert(key, Cell::new(x, y, z_top));
        }
        let old_end = Cell::new(x, y, z);
        let mut extended = false;
        for pth in paths.iter_mut() {
            if *pth.last().unwrap() == old_end {
                pth.extend_from_slice(&seg[1..]);
                extended = true;
                break;
            }
            if pth[0] == old_end {
                let mut np: Path = seg[1..].iter().rev().copied().collect();
                np.extend_from_slice(pth);
                *pth = np;
                extended = true;
                break;
            }
        }
        if !extended {
            paths.push(seg.clone());
        }
        occ.extend(seg.iter().copied());
    }
    state.paths = paths;
    state.occupied = occ;

    for &key in &frontier {
        let tr = state.idle_h_track[&key].clone();
        let st = match state.typ.get(&tr.origin) {
            Some(4) | Some(5) => 0u8,
            Some(t) => *t as u8,
            None => 0,
        };
        let (mut ct, ld) = edge_tracer(&reversed(&tr.path), state.ori[&tr.origin], st);
        if state.htable.needs_flip_to_end(tr.origin) {
            ct = 1 - ct;
        }
        state.ori.insert(key, ori_map(ld, ct, 0));
        state.typ.insert(key, 0);
    }
}
